web/src/navbar_alerts.js

```diff
@@ -70,6 +72,28 @@ export function should_show_server_upgrade_notification(ls) {
     return Date.now() > upgrade_nag_dismissal_duration;
 }
 
+export function maybe_show_empty_required_profile_fields_alert() {
+    const $navbar_alert = $("#navbar_alerts_wrapper").children(".alert").first();
+    if (
+        !$navbar_alert?.length ||
+        $navbar_alert.is("#empty-required-profile-fields-warning") ||
+        $navbar_alert.is(":hidden")
+    ) {
```

Comment:
Can you talk about why this is the correct check? I don't really see why you're looking at `:hidden`, for example.

Reply:
This is used to check if there are any other alerts present in the navbar. When we close the already present alerts they get hidden. So ```:hidden``` check is for adding the warning banner when already present alerts are closed. The ```$navbar_alert.is("#empty-required-profile-fields-warning")``` condition is used to remove the warning banner when the user fills all the required fields.

Reply:
OK. This seems to pretty weirdly misuse the APIs for this module. I made the following changes, which seem to produce a much cleaner result with less code duplication:

``` diff
diff --git a/web/src/navbar_alerts.js b/web/src/navbar_alerts.js
index f47ac04fcb..13cb67ea68 100644
--- a/web/src/navbar_alerts.js
+++ b/web/src/navbar_alerts.js
@@ -73,24 +73,26 @@ export function should_show_server_upgrade_notification(ls) {
 }
 
 export function maybe_show_empty_required_profile_fields_alert() {
+    const empty_required_profile_fields_exist = realm.custom_profile_fields
+          .map((f) => ({
+              ...f,
+              value: people.my_custom_profile_data(f.id)?.value,
+          }))
+          .find((f) => f.required && !f.value);
+    if (!empty_required_profile_fields_exist) {
+        return;
+    }
+
     const $navbar_alert = $("#navbar_alerts_wrapper").children(".alert").first();
     if (
         !$navbar_alert?.length ||
         $navbar_alert.is("#empty-required-profile-fields-warning") ||
         $navbar_alert.is(":hidden")
     ) {
-        const empty_required_profile_fields_exist = realm.custom_profile_fields
-            .map((f) => ({
-                ...f,
-                value: people.my_custom_profile_data(f.id)?.value,
-            }))
-            .find((f) => f.required && !f.value);
-
-        $("#empty-required-profile-fields-warning").remove();
-        if (empty_required_profile_fields_exist) {
-            $("#navbar_alerts_wrapper").html(render_empty_required_profile_fields());
-        }
-        $(window).trigger("resize");
+        open({
+            data_process: "profile-missing-required",
+            rendered_alert_content_html: render_empty_required_profile_fields(),
+        });
     }
 }
 
diff --git a/web/templates/navbar_alerts/empty_required_profile_fields.hbs b/web/templates/navbar_alerts/empty_required_profile_fields.hbs
index 8c075fcdcb..a74eafa37e 100644
--- a/web/templates/navbar_alerts/empty_required_profile_fields.hbs
+++ b/web/templates/navbar_alerts/empty_required_profile_fields.hbs
@@ -1,4 +1,4 @@
-<div id="empty-required-profile-fields-warning" class="alert alert-info">
+<div data-step="1">
     <span>
         {{t 'Your profile is missing required fields.'}}&nbsp;
         <a class="alert-link" href="#settings/profile">{{t 'Edit your profile'}}</a>
```